### PR TITLE
Ports pollution from Skyrat (Ported by Gandalf2k15 from Horizon and made by Azarak), Adds Turf Fires for it

### PR DESCRIPTION
**Pull request by tralezab**
## and that means it wasn't made by me, the mystical and anomalous tralezab either!!

Original PR description from Horizon:

Smoking cigarettes makes either smoke or cannabis smoke (if drugs are smoked)
Smoking vapes makes thicker, but more pleasantly flavored smoke
bonfires make smoke and deadly carbon pollution, which can be avoided by laying down to a degree
Grilling foods emitts related pollutants to them
Chemical reactions can make pollutants
Added perfumes which you can find in loadout
The bartender starts with candle packs, and can purchase scented candle packs in his drobemat
Waste and garbage objects make an unpleasant scent
Spilling blood creates a subtle metallic scent
Lighting matches creates sulphur scent
Drills create thick dust as they drill
Vended tea, coffee and hot choco smell like it

Pollution dissipates slowly over time, twice as fast on planetary turfs

Admins have a Spawn-Pollution verb for bussing

### Turf Fires and Flamethrowers
Turfs can now be ignited and will slowly burn, igniting anyone on the tile.

There's also subtypes of magical turf fires, for mapping and adminbussery, magical fires wont burn out, or wont affect atmos

Hotspots (fire blocks) now have 30% chance to leave a turf fire, creating a cool effect where fiery explosions leave short residual flames around.

Janitor starts with air refreshers in his locker
Reagents now have accelerant quality variable

Flamethrowers have beek reworked, they no longer use atmos or a plasma tank, but instead they use any open container (beakers, bottles, drinking glasses). And they require an accelerant reagent to properly spew fire (fuel, ethanol, napalm etc,). They will now spew hotspots and turf fires up to a range of 4 per each spew. The strength of those depends on how strong is your fuel.

Fire alarms upper trigger temperature changed to 80 from 200, lower from -3 to -10. Fire alarms will also trigger immediately when they are on a lit turf

### how it ties into our funny atmos simulator

scrubbers remove it from the air

image

## Why It's Good For The Game

The addition of smells makes the station much more wonderful. The smells of the kitchen, the janitor being able to febreeze areas, maintenance smelling horrible, etc.

The addition of smoke makes explosions, fires, and more much more immersive. See below:

image

## Changelog
:cl:
add: The sense of smell - Ported from Skyrat, which is ported from Horizon. God, I love open contribution!
add: Added a pollution system, which is used for air pollutants, such as cigarette smoke, deadly carbon pollution, perfumes, food scents etc.
/:cl:

### CODE REVIEW

**Comment by Ghilker** [comment]
Bad idea, now plasmafires create smokes that impedes the vision and @LemonInTheDark doesn't want that, also this

unknown-19

**Comment by AMonkeyThatCodes** [comment]
if lemon in the dark doesn't want that I'm sure he'll post his concerns himself.

File: code/datums/components/grillable.dm
Comment (AMonkeyThatCodes) [inline code/datums/components/grillable.dm:20]: ok arm

Comment (tralezab) [reply code/datums/components/grillable.dm]: kek

**Comment by tralezab** [comment]
i checked myself to see how skyrat looks purdier but maybe it was just a josh/colored lighting because the icons for smoke are the same over there as they are over here

**Comment by Dorsisdwarf** [comment]
I mean, unless it looks like that to players just give ghosts "toggle smoke" to turn it off locally

**Comment by TemporalOroboros** [comment]
You know, I was thinking of giving the chaplain incense at some point. I'm in favor honestly.

**Comment by oklews** [comment]
Thank god, I've always wanted the second hand smoking realism

**Comment by KittyNoodle** [comment]
I like this, gives chemists and Janitors more to do, but the alpha values should be turned down, maybe half of what they are now?

**Comment by AMonkeyThatCodes** [comment]
just try it like this first.

**Comment by KittyNoodle** [comment]
Are different races gonna have different smelling abilities like how tongues have different taste sensitivities (eg, lizards have high smell, ethereals have no smell to act with their no taste)

**Comment by CRITAWAKETS** [comment]
Why isn't this an atmos job?

**Comment by LemonInTheDark** [comment]
@tralezab could you describe to me how pollution is meant to work/move, and what might make it expensive?

Outside of that, I don't like the fully transparent smoke, I kinda like how plasma fires look currently, though the smoothing and distraction from the tiling would be nice. That and I think being on fire is a big enough problem already, not being able to see anything at all is too far.

**Comment by MrStonedOne** [comment]
> if your on fire i think you have bigger problems than not being able to see

This isn't an argument:

"you have bigger problems" Which would be exactly why you shouldn't make it worse by obfuscating their ability to see water tanks, fire extinguishers on the ground, or other shit.

> i dont think smoke obscures the walls/doors so you should be able to find the exit well enough

Not everything that can block your movement is a turf or door and trying to walk around something you can't see would absolutely be an issue.

**Review by MrsTonedOne — CHANGES_REQUESTED** [review]
:-1: why are we creating a secondary atmos system for smoke spreading and dissipation effects?

**Review by ghost — APPROVED** [review]
why are you using the review system for this

**Comment by MrStonedOne** [comment]
> why are you using the review system for this

because i reviewed the code and do not approve of it for the reasons listed in the review comment.

**Comment by AMonkeyThatCodes** [comment]
@tralezab you should remove the old atmos system to adress MSO's complaint

**Comment by AlinaStarkova** [comment]
> 
> 
> > if your on fire i think you have bigger problems than not being able to see
> 
> This isn't an argument:
> 
> "you have bigger problems" Which would be exactly why you shouldn't make it worse by obfuscating their ability to see water tanks, fire extinguishers on the ground, or other shit.

not my point, the smoke doesnt get opaque until its REALLY bad, fires dont block vision even with the smoke they put off until its burning so badly that you will literally die in seconds, and anyone not in the fire also will be able to safely grab water

again, this has existed for awhile on other servers and never had an issue with this, and those servers are HRP, aka "things that make us die easily are BAD!!!!!" if this made fires more deadly, it wouldnt exist

**Comment by MrStonedOne** [comment]
HRP also means "setting things on fire for no reason is bad" and would see less abuse of a system like this then LRP.

Try again.

**Comment by AlinaStarkova** [comment]
> 
> 
> HRP also means "setting things on fire for no reason is bad" and would see less abuse of a system like this then LRP.
> 
> Try again.

you are aware the maintainers and such are actively testmerging this right? and so far see no issues not even with people actively trying to pollute the station yes? you can hop onto a round to see for yourself to see how bad it gets (and keep in mind, they are trying HARD to pollute as well) 

your entire disagreement stems from a single image and your assumption about how it works in practice, why not go see it in practice first?

**Comment by LemonInTheDark** [comment]
I didn't ask for this to be testmerged. I've seen it in a plasma fire on a local though, and it's not something I'd want I don't think.

During normal gameplay it seems fine, close to a non issue. Mostly because not a whole lot of stuff makes pollution in the first place

**Comment by AlinaStarkova** [comment]
> 
> 
> I didn't ask for this to be testmerged. I've seen it in a plasma fire on a local though, and it's not something I'd want I don't think.
> 
> During normal gameplay it seems fine, close to a non issue. Mostly because not a whole lot of stuff makes pollution in the first place

arm did, iirc it was testmerged on sybil

File: code/modules/atmospherics/environmental/LINDA_fire.dm
Comment (Azarak) [inline code/modules/atmospherics/environmental/LINDA_fire.dm:132]: The way it works on horizon is that hotspots don't make any pollution, but they can create(30%) a residual turf fire when they dissipate, which then emitts pollution in a more controlled and slow way (as highlighted here https://user-images.githubusercontent.com/6381979/126042078-cd21a363-2c08-4e2c-8bcd-f36e8dfd53fc.png the entire room was on fire). What's done here is a really excessive emission, I'd suggest changing that

**Comment by Azarak** [comment]
> 
> 
> 👎 why are we creating a secondary atmos system for smoke spreading and dissipation effects?

It's nothing like an atmospherics subsystem, it uses a special way of processing (for both gameplay and computing) and it handles much less things and it's 100x (not even an exaggeration) times lighter.

**Comment by tralezab** [comment]
Well, the port is uuuh... done, so it's really just down to maintainers to decide

**Comment by AMonkeyThatCodes** [comment]
fix the things azarak mentioned noob

**Comment by tralezab** [comment]
ogey :)

**Comment by AMonkeyThatCodes** [comment]
you should either port it or remove smoke from the hotspots.

**Comment by tralezab** [comment]
yeah doesn't removing smoke from hotspots fix a lot of issues

**Comment by AMonkeyThatCodes** [comment]
you put them on hotspots, you should just put them on the turf fires and port those too, theres no reason not to. It looks good.

File: code/modules/pollution/pollution_datum.dm
Comment (Sinestia) [inline code/modules/pollution/pollution_datum.dm:43]: dominant

File: code/controllers/subsystem/pollutant_subsystem.dm
Comment (Sinestia) [inline code/controllers/subsystem/pollutant_subsystem.dm:15]: dissipation

**Review by Sinestia — CHANGES_REQUESTED** [review]
just mentioning a couple var names

**Comment by LemonInTheDark** [comment]
> 
> 
> > 👎 why are we creating a secondary atmos system for smoke spreading and dissipation effects?
> 
> It's nothing like an atmospherics subsystem, it uses a special way of processing (for both gameplay and computing) and it handles much less things and it's 100x (not even an exaggeration) times lighter.

100x is maybe a bit much, though I guess it depends on how you think about things.
I did some plasma fire tests on a local with the smoke thing arm has currently, which is perhaps a bit out of band. pollution took about 1/6th the cpu time as atmos. That doesn't tell the full story cause one of these subsystems was taking like 10 seconds to fully fire and the other wasn't, but it still bothers me.

image
Funnily enough it made overtime in about the same pattern as ssair.
When an excited group equalizes, we need to sum the contents of all connected tiles and equalize.

From reading the code here you're doing something similar, but only with connected tiles that haven't been touched?

It generates a weird amount of overtime, especially because it's not consistent at all. It spikes up to like 30% or so and then steps down to 0, spikes again. Makes me think there's a big equalization step but I can't find anything of that description.

image

The ProcessCell proc eats a good amount of cpu time at load, but it's kinda hard to tell what parts are actually causing problems because it's all one proc.

image

I'm not going to comment on how a large buildup of smoke (maybe 70 height) on a tile looks, how long it takes to dissipate, any of that. I think it'd be unfair to you since by the sounds of it it doesn't match your planned implementation.

I don't like how it looks over space, or how it looks put on top of the sm.

Ahhhh lemme see. The not being able to see due to fires thing needs to go. I unlike codemonkey also don't like how a full on smogged burn looks. Not that hotspots are any great shakes currently, but at least they're clear.

Outside of that, there's a lot of stuff here that's trying to solve atmos like problems, but in like, less impactful ways? It absolutely dies under load because of that. Stuff like assoc lists instead of turf vars. At least I assume that's why I saw the spikes in overtime that I did, not sure what else would cause that. Dissipation doesn't look expensive so I shouldn't think it's that.

https://github.com/tgstation/tgstation/blob/e7a44ff111b5b16da66f106783a850a2721a6089/code/__DEFINES/pollution.dm#L1-L6

I am a bit worried about this post reflection. It seems a good bit like atmos but again and less intensive. Not as complex, but atmos isn't complex for no reason, and some of this stuff seems like it could use the complexity. Maybe my concept of what "load" means for pollution is fucked, and that's why I saw what I did. Can't tell.

https://github.com/tgstation/tgstation/blob/e7a44ff111b5b16da66f106783a850a2721a6089/code/modules/pollution/pollution_datum.dm#L140-L144

That's partially why I asked tralezab what he knew about the system. I wanted to see if he understood it past just porting it over, and if he had a good idea of what it does well/poorly

I don't know how I feel about this on the whole, but I'm apprehensive about it. Not because the system's bad at the core, I understand why you're doing what you're doing, I think I might do the same in some cases.

It's just like, it's another flow like system, and it runs into a lot of the same problems, especially if we misuse it, which I don't trust us not to do. When @AMonkeyThatCodes and @tralezab first asked me about this system I told them that they could port it if they like, did maintainer reviews and shit. But I'm realizing that this is somewhat complex, and I'm not sure I can hold to that anymore. Having a harder time justifying it to myself

**Comment by Azarak** [comment]
@LemonInTheDark Those profilings arent accurate because pollution would run with a priority over atmos(1/4th times fire rate), which then would run so hard it'd throttle itself even before doing a fraction of what it wants to. I said 100x in a perspective of how gas cell shares with another gas cell, in comparison to pollution sharing with pollution.

There's probably a good amount of optimizing that can be done here, as I wasn't aware that setting an associative key was so expensive until today.

And yeah hotspot fires making smog is not a thing on our fork, nor it should be here either IMO

**Comment by LemonInTheDark** [comment]
> 
> 
> @LemonInTheDark Those profilings arent accurate because pollution would run with a priority over atmos(1/4th times fire rate), which then would run so hard it'd throttle itself even before doing a fraction of what it wants to. I said 100x in a perspective of how gas cell shares with another gas cell, in comparison to pollution sharing with pollution.

I think this number just scales poorly with turfs processing for pollution. In my profiles from the tower post it's like 4x faster then turf atmos, in terms of total cpu per call, which feels a bit too small.

> There's probably a good amount of optimizing that can be done here, as I wasn't aware that setting an associative key was so expensive until today.

It's not like, expensive. From what I understand it's a lot cheaper then |=, it's just not as cheap as a var lookup. Like I said it makes sense to not use say, the archive system, or turf vars for excitement and all, it makes the code and the footprint of the code a lot nicer, but they're there instead of an assoc lookup for a reason. It may just be this bad at high load, but I don't trust our feature coders not to go to high load.

> And yeah hotspot fires making smog is not a thing on our fork, nor it should be here either IMO

Ye

**Comment by Azarak** [comment]
>I think this number just scales poorly with turfs processing for pollution. In my profiles from the tower post it's like 4x faster then turf atmos, in terms of total cpu per call, which feels a bit too small.
Thing is, gas sharing takes a good amount of time and calls to actually equalize and become unactive, it happens with a fast fire rate too, while pollution is designed to equalize very quickly, so there's much less processing left, meaning you'll have a drastically lower amount of calls of pollution's ProcessCell compared to atmos for similar tasks.
In your profiling, what you ran into is a case where the pollution was spawned repeatedly all over without having a chance to settle down, which the issue was highlighted even more so by it having a higher priority

**Comment by LemonInTheDark** [comment]
Yeah. Due to hotspots making pollution on process I'd assume.

**Comment by EricZilla** [comment]
God I hope this opens the door to having like, fog. Imagine the space dragon portal producing a thin but slowly expanding fog

**Comment by tralezab** [comment]
we were talking about xeno eggs producing a fog which would be dope

**Comment by Azarak** [comment]
If you modify the appearance code of pollution you could end up with very advanced and customizable visual stuff, various overrides, planes, filters etc. (particles? i dunno how they work yet)

**Comment by GoldenAlpharex** [comment]
From experiencing it on Skyrat, I definitely agree about smoke not being produced by hotspots.

**Comment by AMonkeyThatCodes** [comment]
just port the damn turf fires i beg you.

**Comment by tralezab** [comment]
working on turf fires right now

**Comment by AMonkeyThatCodes** [comment]
👑

**Comment by AMonkeyThatCodes** [comment]
u dropped that king

**Comment by tralezab** [comment]
image

**Comment by tralezab** [comment]
what the f- that was a gif a second ago

**Comment by Caldony** [comment]
finally my pyre sect can be turned into "the industrial revolution and its consequences" sect

**Comment by AMonkeyThatCodes** [comment]
i dont think you should be adding new sounds to the flamethrower in a pr for pollution. thats some beestation port type stuff ;p

**Review by MrsTonedOne — CHANGES_REQUESTED** [review]
Still blocking this until you find a way to **CAP** the max opacity so these do not block vision.

With how unbalanced new features tend to be on this codebase we can revisit making these able to fully block out vision of the tiles later, once things have settled down balance and new feature abuse wise.

**Comment by MrStonedOne** [comment]
image
image

**Comment by ExcessiveUseOfCobblestone** [comment]
Your PR is missing proper format, an inappropriate title, or contains incorrect/missing information in your PR body.

Please look over your PR and make sure all items are filled out in a way that someone looking back years from now can easily access and understand.

Why its good for the game

________

Moreover I am not a fan of "it just goes away on its own", there should be some atmos-related aspect to it that allows for smoke to dissapate faster.

If these are going to block vision, are there any mitigatable factors?

**Comment by AMonkeyThatCodes** [comment]
define mitigatable factors

**Comment by GoldenAlpharex** [comment]
> Moreover I am not a fan of "it just goes away on its own", there should be some atmos-related aspect to it that allows for smoke to dissapate faster.
> 
> If these are going to block vision, are there any mitigatable factors?

There is, actually! Using a portable scrubber speeds up the pollution removal by a major margin. We're talking... At least 3 to 5 times faster, if not more, I don't remember the exact numbers.

**Comment by LemonInTheDark** [comment]
> 
> 
> Still blocking this until you find a way to **CAP** the max opacity so these do not block vision.
> 
> With how unbalanced new features tend to be on this codebase we can revisit making these able to fully block out vision of the tiles later, once things have settled down balance and new feature abuse wise.

Bruh this is my feature to block. Can't have shit on tg

**Comment by AMonkeyThatCodes** [comment]
image
that got fixed now btw

File: code/controllers/subsystem/pollutant_subsystem.dm
Comment (Watermelon914) [reply code/controllers/subsystem/pollutant_subsystem.dm:1]: also put the priority behind atmospherics so atmos runs first.

**Comment by tralezab** [comment]
still have no idea why turf fires hard del :sadcat:

**Comment by LemonInTheDark** [comment]
does the ref searcher find anything?

**Comment by tralezab** [comment]
hard dels gone!

**Comment by FlamingLily** [comment]
> ## This was not made by myself. Azarak made it and you can find it at the original PR: [hrzntal/horizon#468](https://github.com/hrzntal/horizon/pull/468)
> Original PR

> Fixes Pollution Emission Caps


no bro.

https://github.com/hrzntal/horizon/pull/454

**Comment by tralezab** [comment]
@FlamingLily apologies, was mislinked on the skyrat pr. shoulda double checked

**Comment by FlamingLily** [comment]
> 
> 
> @FlamingLily apologies, was mislinked on the skyrat pr. shoulda double checked

man.... I told him too, smh my head

**Comment by LemonInTheDark** [comment]
> 
> 
> @tralezab could you describe to me how pollution is meant to work/move, and what might make it expensive?

I'm assuming this just got missed. Still really want to know how much you know about this system and what it's doing.

File: code/controllers/subsystem/turf_fire_subsystem.dm
Comment (MrStonedOne) [inline code/controllers/subsystem/turf_fire_subsystem.dm:9]: This needs statistics.

If you make it a processing subsystem and use the macros, you'll get the statistics for free.

Comment (LemonInTheDark) [reply code/controllers/subsystem/turf_fire_subsystem.dm:9]: it also needs tick checking

**Review by Kylerace — CHANGES_REQUESTED** [review]
coming by to point out attainable improvements. this code is REALLY trigger happy with assoc sets in hot areas which are pretty slow, heres a benchmark i did thats biased for the assoc test since its Cut() ing the list less
[benchmark.txt](https://github.com/tgstation/tgstation/files/7175258/benchmark.txt)
Screenshot_1301

File: code/modules/pollution/pollution_datum.dm
Comment (LemonInTheDark) [reply code/modules/pollution/pollution_datum.dm:142]: assoc setting for large lists should be faster then |=, at least that's the urban legend, that's why it's done here

Comment (Kylerace) [reply code/modules/pollution/pollution_datum.dm:142]: from my first benchmark that threshold is somewhere above 100,000

File: code/modules/pollution/pollution_datum.dm
Comment (Kylerace) [reply code/modules/pollution/pollution_datum.dm:132]: a lot of these procs need sensible spacing

File: code/modules/pollution/pollution_datum.dm
Comment (LemonInTheDark) [reply code/modules/pollution/pollution_datum.dm:142]: frankly this list just shouldn't exist, but it'd mean adding a turf var to be properly consistent

Comment (Kylerace) [reply code/modules/pollution/pollution_datum.dm:142]: im talking about sharing_turfs here

Comment (LemonInTheDark) [reply code/modules/pollution/pollution_datum.dm:142]: ok yeah that's just silly

File: code/game/objects/effects/turf_fire.dm
Comment (Kylerace) [inline code/game/objects/effects/turf_fire.dm:131]: use as anything here

File: code/game/objects/effects/turf_fire.dm
Comment (Kylerace) [inline code/game/objects/effects/turf_fire.dm:133]: why do you have two instances of 
```
if(!magical)
```
?

File: code/game/objects/effects/turf_fire.dm
Comment (Kylerace) [inline code/game/objects/effects/turf_fire.dm:63]: needless assoc list again, fires is never searched only iterated through

**Comment by LemonInTheDark** [comment]
Aight.
So the code currently has some problems. I'm not closing it because of that though. Those are fixable.
I'm closing it because this system is drastically similar to atmospherics flow, which is a reasonably complex thing to maintain. It's only workable at this point because I've documented the intent of everything, but it's still teetering on that edge.

The code here isn't nearly as complex, mostly because it can trade some processing time for cleanliness. It has some core flaws in how the flow works that really bother me, like, it's not consistent at all, movement depends on the order things were initialized in. It might not be a problem for everyone, but that's not something I can reasonably let in. 
Attempting to fix something like that necessitates a lot of complexity to work well, and just puts us in a bad spot.

I've spent months of my time fixing complex systems like that, and I don't want to put that on anyone else down the line in a year or three when I die in mysterious circumstances. 

I gave my signoff on this being worked on in the first place. Shouldn't have done that, nothing you could have done would have changed my mind. I'm sorry for not realizing that at the time, and for dragging this out so long. Your work is not unappreciated.

o7

**Pull request closed without merging**
